src/features/code_actions.zig

```diff
@@ -366,6 +382,272 @@ fn handleVariableNeverMutated(builder: *Builder, actions: *std.ArrayListUnmanage
+    pub fn getSortSlice(self: ImportDecl) []const u8 {
+        switch (self.getKind()) {
+            .file => {
+                if (std.mem.indexOfScalar(u8, self.getSortValue(), '/') != null) {
+                    return self.value[1 .. self.getSortValue().len - 1];
+                }
+                return self.getSortName();
+            },
+            // There used to be unreachable for other than file and package, but the user
+            // can just write @import("std") twice.
+            else => return self.getSortName(),
+        }
+    }
+
+    pub fn getSortName(self: ImportDecl) []const u8 {
+        return self.parent_name orelse self.name;
+    }
+
+    pub fn getSortValue(self: ImportDecl) []const u8 {
+        return self.parent_value orelse self.value;
+    }
+
+    /// returns true if there should be an empty line between these two imports
+    /// assumes `lessThan(void, lhs, rhs) == true`
+    pub fn addSeperator(lhs: ImportDecl, rhs: ImportDecl) bool {
+        const lhs_kind = @intFromEnum(lhs.getKind());
+        const rhs_kind = @intFromEnum(rhs.getKind());
+        if (rhs_kind <= @intFromEnum(Kind.build_options)) return false;
+        return lhs_kind != rhs_kind;
+    }
+
+    pub fn getSourceStartIndex(self: ImportDecl, tree: Ast) usize {
+        return offsets.tokenToIndex(tree, self.first_comment_token orelse tree.firstToken(self.var_decl));
+    }
+
+    pub fn getSourceEndIndex(self: ImportDecl, tree: Ast, include_line_break: bool) usize {
+        const token_tags = tree.tokens.items(.tag);
+
+        var last_token = ast.lastToken(tree, self.var_decl);
+        if (last_token + 1 < tree.tokens.len - 1 and token_tags[last_token + 1] == .semicolon) {
+            last_token += 1;
+        }
+
+        const end = offsets.tokenToLoc(tree, last_token).end;
+        if (!include_line_break) return end;
+        return std.mem.indexOfNonePos(u8, tree.source, end, &.{ ' ', '\t', '\n' }) orelse tree.source.len;
+    }
+
+    /// similar to `offsets.nodeToLoc` but will also include preceding comments and postfix semicolon and line break
+    pub fn getLoc(self: ImportDecl, tree: Ast, include_line_break: bool) offsets.Loc {
+        return .{
+            .start = self.getSourceStartIndex(tree),
+            .end = self.getSourceEndIndex(tree, include_line_break),
+        };
+    }
+};
+
+pub fn getImportsDecls(builder: *Builder, allocator: std.mem.Allocator) error{OutOfMemory}!std.ArrayListUnmanaged(ImportDecl) {
+    const tree = builder.handle.tree;
+    var imports = std.ArrayListUnmanaged(ImportDecl){};
+    errdefer imports.deinit(allocator);
+
+    const node_tags = tree.nodes.items(.tag);
+    const node_data = tree.nodes.items(.data);
+    const node_tokens = tree.nodes.items(.main_token);
+
+    // iterate until no more imports are found
+    var updated = true;
+    while (updated) {
+        updated = false;
+        next_decl: for (tree.rootDecls()) |node| {
```

Comment:
I tried to come up with something smarter, but couldn't. The problem is that they can all depend on each other like a chain:

```zig
const Rational = big.rational; // Resolved in iteration #4
const big = math.big;          // Resolved in iteration #3
const math = std.math;         // Resolved in iteration #2
const std = @import("std");    // Resolved in iteration #1
```

But Zig compiler has to deal with this somehow, so it might be worth it to take inspiration there.

I could attempt to improve it in a followup PR.